### ISSUE TITLE
Crash when uploading timeout is hit

### ISSUE THREAD

**Issue by daxpedda** [getsentry/sentry-native#322]
**Description**
Crash when capturing too many events in a short period of time.

**When does the problem happen**
- [ ] During build
- [x] During run-time
- [ ] When capturing a hard crash

**Environment**
- OS: Windows 10, 64-bit
- Compiler: MSVC 19
- CMake config: BUILD_SHARED_LIBS=OFF

**Steps To Reproduce**
```c
sentry_options_t *options = sentry_options_new();
sentry_init(options);

for (i = 0; i < 3; ++i)
{
    sentry_value_t event = sentry_value_new_event();
    sentry_capture_event(event);
}

sentry_shutdown();
```

```
Assertion failed: !bgw->running && bgw->task_count == 0, file C:\rust\sentry-contrib-native\sentry-contrib-native-sys\sentry-native\src\sentry_sync.c, line 163
```

Sometimes this is also added to the error message:
```
[27584:1256:20200626,225109.609:ERROR file_io_win.cc:161] CreateFile C:\rust\sentry-contrib-native\target\debug\build\sentry-contrib-native-231a7bbe253c3ca2\out\.sentry-native\e39cefa0-2755-4f41-d9db-c2693faaf780.run\__sentry-breadcrumb1: The system cannot find the path specified. (3)
[27584:1256:20200626,225109.611:ERROR file_io_win.cc:161] CreateFile C:\rust\sentry-contrib-native\target\debug\build\sentry-contrib-native-231a7bbe253c3ca2\out\.sentry-native\e39cefa0-2755-4f41-d9db-c2693faaf780.run\__sentry-breadcrumb2: The system cannot find the path specified. (3)
[27584:1256:20200626,225109.613:ERROR file_io_win.cc:161] CreateFile C:\rust\sentry-contrib-native\target\debug\build\sentry-contrib-native-231a7bbe253c3ca2\out\.sentry-native\e39cefa0-2755-4f41-d9db-c2693faaf780.run\__sentry-event: The system cannot find the path specified. (3)
```

~~I need an average of 10 tries to make it crash.~~

The error seems to be triggered the easiest when there is a big delay between the client and server.
Additionally I was not able to reproduce this error anywhere else than on Windows, I tried both Linux and macOS.
I first discovered it through Github Actions, so this can be tested for.

**Comment by daxpedda** [comment]
I noticed that I had trouble reproducing this on my local onpremise server. I therefore concluded that the issue must be timing related.

**Comment by daxpedda** [comment]
I used `netem` to fake a 2000ms ping, which triggered this crash on every single try, so there is definitely a relation to timing here.

**Comment by Swatinem** [comment]
So we do have the 2s timeout on shutdown, to flush all unsent envelopes. The assertion is hit when we run into that timeout.

The errors you see logged are coming from crashpad I think, which is still attached and configured with attachments that we clean up before trying to flush the transport.

**Comment by daxpedda** [comment]
Oh, I didn't realize this was intentional.
Is there an interest to change this behaviour? I don't think the application should crash just because it hit the time-out, maybe set a queryable flag somewhere or return an error code from `shutdown`?

**Comment by Swatinem** [comment]
This should be fixed now with #323
